### PR TITLE
Adding support for Microsoft.NET.Sdk.Worker

### PR DESCRIPTION
**Pull request by jmprieur**
# Adding support for Microsoft.NET.Sdk.Worker

This PR:
- Fixes an issue with Downstream API. The service could be registered several times and systematically as a scoped service, independently of the token acquisition.
- Adds a worker sample using IDownstreamApi.

Fixes:
- #1755
- #2636
- #2515
- #2212

Related to:
- #1124

cc: @DavidParks8, @jennyf19

### CODE REVIEW

File: tests/DevApps/ContosoWorker/Program.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+using ContosoWorker;
+using Microsoft.Identity.Abstractions;
+using Microsoft.Identity.Web;
+using Microsoft.Identity.Web.TokenCacheProviders.InMemory;
+
+var builder = Host.CreateApplicationBuilder(args);
+
+builder.Services.AddHostedService<Worker>();
+
+builder.Services.AddTokenAcquisition(isTokenAcquisitionSingleton:true)
```
Comment (jmprieur) [inline tests/DevApps/ContosoWorker/Program.cs:10]: We could introduce a:
TokenAcquirerFactory.GetDefault(hostBuilder) adds the token acquisition, the cache and sets the MicrosoftIdentityOptions.

Note that the AddTokenAcquisition in this case uses an Sdk host, which means it's not drawing ASP.NET Core services.

File: tests/DevApps/ContosoWorker/Worker.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+using Microsoft.Identity.Abstractions;
+
+namespace ContosoWorker;
+
+public class Worker : BackgroundService
+{
+    private readonly ILogger<Worker> _logger;
+    private readonly IDownstreamApi _downstreamApi;
+
+    public Worker(ILogger<Worker> logger, IDownstreamApi downstreamApi)
```
Comment (jmprieur) [inline tests/DevApps/ContosoWorker/Worker.cs:10]: Just inject the IDownstreamApi or GraphServiceClient etc ...

File: tests/DevApps/ContosoWorker/Worker.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        while (!stoppingToken.IsCancellationRequested)
+        {
+            if (_logger.IsEnabled(LogLevel.Information))
+            {
+                _logger.LogInformation("Worker running at: {time}", DateTimeOffset.Now);
+                var result = await _downstreamApi.CallApiAsync("MyWebApi");
```
Comment (jmprieur) [inline tests/DevApps/ContosoWorker/Worker.cs:23]: and use it

File: tests/DevApps/ContosoWorker/appsettings.json [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        {
+          "SourceType": "KeyVault",
+          "KeyVaultUrl": "https://webappsapistests.vault.azure.net",
+          "KeyVaultCertificateName": "Self-Signed-5-5-22"
+        }
+      ]
+    },
+
+  "MyWebApi": {
+    "BaseUrl": "https://graph.microsoft.com/v1.0",
```
Comment (jmprieur) [inline tests/DevApps/ContosoWorker/appsettings.json:16]: Used Microsoft Graph as a well-known service in this sample.

File: src/Microsoft.Identity.Web.DownstreamApi/DownstreamApiExtensions.cs [hunk 5]
```diff
@@ -52,14 +54,50 @@ public static class DownstreamApiExtensions
...
-
-            // https://docs.microsoft.com/en-us/dotnet/standard/microservices-architecture/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests
-            services.AddScoped<IDownstreamApi, DownstreamApi>();
             services.Configure(serviceName, configureOptions);
+            RegisterDownstreamApi(services);
+
             return services;
         }
 
+        private static void RegisterDownstreamApi(IServiceCollection services)
```
Comment (bgavrilMS) [inline src/Microsoft.Identity.Web.DownstreamApi/DownstreamApiExtensions.cs:63]: Is it not possible to unit test this? Seems like you just need to register a few mock ITokenAcquisition / IDownstream APIs with different lifetimes, get them from the service collection and assert on their lifetime.

Comment (jmprieur) [reply src/Microsoft.Identity.Web.DownstreamApi/DownstreamApiExtensions.cs]: Done. Thanks

**Comment by jforward5** [comment]
How long before this is available?

**Comment by JoshLozensky** [comment]
This went out in Feb 2024 version 2.17.0

**Pull request merged**